File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> result = new ArrayList<>();
        
        // Handle edge case where pattern is longer than the string
        if (s == null || p == null || s.length() < p.length()) {
            return result;
        }

        // Frequency arrays for 26 lowercase English letters
        int[] pCounts = new int[26];
        int[] sCounts = new int[26];

        // Fill the reference pattern frequency map
        for (int i = 0; i < p.length(); i++) {
            pCounts[p.charAt(i) - 'a']++;
        }

        int pLen = p.length();
        int sLen = s.length();

        // Slide the window across string s
        for (int i = 0; i < sLen; i++) {
            // Add the rightmost character to the current window
            sCounts[s.charAt(i) - 'a']++;

            // Remove the leftmost character that falls out of the window
            if (i >= pLen) {
                sCounts[s.charAt(i - pLen) - 'a']--;
            }

            // If the window matches the pattern length, compare frequencies
            if (i >= pLen - 1) {
                if (matches(pCounts, sCounts)) {
                    result.add(i - pLen + 1);
                }
            }
        }

        return result;
    }

    // Helper method to compare fixed-size arrays in O(1) time
    private boolean matches(int[] pCounts, int[] sCounts) {
        for (int i = 0; i < 26; i++) {
            if (pCounts[i] != sCounts[i]) {
                return false;
            }
        }
        return true;
    }
}

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> list[int]:
        if len(p) > len(s):
            return []
        
        # Initialize frequency arrays for 26 lowercase English letters
        p_counts = [0] * 26
        s_counts = [0] * 26
        
        # Populate the pattern frequencies
        for char in p:
            p_counts[ord(char) - ord('a')] += 1
            
        p_len = len(p)
        result = []
        
        # Slide the window across string s
        for i in range(len(s)):
            # Add current character to window
            s_counts[ord(s[i]) - ord('a')] += 1
            
            # Remove character that falls out of the window from the left
            if i >= p_len:
                s_counts[ord(s[i - p_len]) - ord('a')] -= 1
                
            # Compare current window to pattern frequencies
            if s_counts == p_counts:
                result.append(i - p_len + 1)
                
        return result
